accepted/engine-host-functions.md

```diff
@@ -0,0 +1,211 @@
+ 
+As an instance is expected to live only for the duration of the request, each instance would be associated with its own `Store` that is dropped when the request completes.
+ 
+Given the nature of `Store` isolation in Wasmtime, this means host function definitions must be recreated for each and every request handled by the service.
+ 
+If instead host functions could be defined for an `Engine` (in addition to `Store`), then the repeated effort to define the host functions would be eliminated and the time it takes to create an instance to handle a request would therefore be noticeably reduced.
+ 
+# Proposal sketch
+[proposal]: #proposal
+ 
+As `Func` is inherently tied to a `Store`, it cannot be used to represent a function associated with an `Engine`.
+ 
+This RFC proposes not having an analogous type to represent these functions; instead, users will define the function via a method on `Engine`, but use `Store` (or `Linker`) to get the `Func` representing the function.
+ 
+## Overview example
+ 
+A simple example that demonstrates defining a host function in an `Engine`:
+ 
+```rust
+let engine = Engine::default();
+engine.define_function("", "hello", |caller: Caller<'_>, ptr: i32, len: i32| {
+    let mem = caller.get_export("memory").unwrap().into_memory().unwrap();
+    println!(
+        "Hello {}!",
+        std::str::from_utf8(unsafe{ &mem.data_unchecked()[ptr..ptr + len] }).unwrap()
+    );
+})?;
+ 
+let module = Module::new(
+    &engine,
+    r#"
+    (module
+        (import "" "hello" (func $hello (param i32 i32)))
+        (memory (export "memory") 1)
+        (func (export "run")
+            i32.const 0
+            i32.const 5
+            call $hello
+        )
+        (data (i32.const 0) "world")
+    )
+    "#,
+)?;
+ 
+// Note: given that `Module` is `Sync + Send`, this could be on a request handling thread
+// As nothing is defined locally in the `Linker`, the engine definitions will
+// be used for satisfying imports
+let store = Store::new(module.engine());
+let linker = Linker::new(&store);
+ 
+let instance = linker.instantiate(&module)?;
+let run = instance
+    .get_export("run")
+    .unwrap()
+    .into_func()
+    .unwrap()
+    .get0::<()>();
+ 
+run()?;
+```
+ 
+## Changes to `wasmtime::Engine`
+ 
+This RFC proposes a `define_function` method be added to `Engine`:
+ 
+```rust
+/// Defines a function for the `Engine`.
+///
+/// Use `Store::get_engine_function` to get a `Func` representing the function.
+///
+/// A `Linker` will automatically default to functions on the associated `Engine`
+/// when resolving a function import that is not defined locally to the linker.
+pub fn define_function<Params, Results>(&self, module: &str, name: &str, func: impl IntoEngineFunc<Params, Results>) -> Result<()>;
```

Comment:
I think we may not end up needing `IntoEngineFunc` as a separate trait here? It's mentioned below that this trait will also be implemented for `Fn(Caller, &[Val], &mut [Val]) -> Result<Trap> + `static` but I don't think we can do that because we can't infer the type signature from the closure. If that's the only reason, though, then this could have two methods, one taking `IntoFunc` and the other being the mirror of `Func::new` which takes a signature and a function?

Reply:
:+1: it would be really nice to avoid another `IntoFunc` style trait.

* They aren't great from a user point of view since they are basically "special" sealed traits and you have to figure out what shape of things are expected.
* This seems, as Alex mentions, basically identical to the existing trait. Would be great to share a single trait if possible.

Reply:
The existing `IntoFunc` method is inherently tied to `Store`, which is why it isn't being used in this context.

I'm not a fan of more boilerplate code, but I'm open to somehow extending `IntoFunc` for this use case.

Regarding `Fn(Caller, &[Val], &mut [Val])`, you're absolutely right that we can't infer the Wasm type of the function with that so another method will be necessary.

Reply:
Ah right that's a good point, but I think we can add new functions as necessary to that trait since it's not intended to be implemented externally. We might have roughtly the same amount of boilerplate internally but  I think we should be able to get away witth one trait.

Reply:
Agreed.  I'll update the draft with extensions for `IntoFunc`.